allenact/embodiedai/mapping/mapping_utils/point_cloud_utils.py

```diff
@@ -0,0 +1,538 @@
+)
+
+
+def camera_space_xyz_to_world_xyz(
+    camera_space_xyzs: torch.Tensor,
+    camera_world_xyz: torch.Tensor,
+    rotation: float,
+    horizon: float,
+) -> torch.Tensor:
+    """Transforms xyz coordinates in the camera's coordinate frame to world-
+    space (global) xyz frame.
+
+    This code has been adapted from https://github.com/devendrachaplot/Neural-SLAM.
+
+    **IMPORTANT:** We use the conventions from the Unity game engine. In particular:
+
+    * A rotation of 0 corresponds to facing north.
+    * Positive rotations correspond to CLOCKWISE rotations. That is a rotation of 90 degrees corresponds
+        to facing east. **THIS IS THE OPPOSITE CONVENTION OF THE ONE GENERALLY USED IN MATHEMATICS.**
+    * When facing NORTH (rotation==0) moving ahead by 1 meter results in the the z coordinate
+        increasing by 1. Moving to the right by 1 meter corresponds to increasing the x coordinate by 1.
+         Finally moving upwards by 1 meter corresponds to increasing the y coordinate by 1.
+         **Having x,z as the ground plane in this way is common in computer graphics but is different than
+         the usual mathematical convention of having z be "up".**
+    * The horizon corresponds to how far below the horizontal the camera is facing. I.e. a horizon
+        of 30 corresponds to the camera being angled downwards at an angle of 30 degrees.
+
+    # Parameters
+    camera_space_xyzs : A 3xN matrix of xyz coordinates in the camera's reference frame.
+        Here `x, y, z = camera_space_xyzs[:, i]` should equal the xyz coordinates for the ith point.
+    camera_world_xyz : The camera's xyz position in the world reference frame.
+    rotation : The world-space rotation (in degrees) of the camera.
+    horizon : The horizon (in degrees) of the camera.
+
+    # Returns
+    3xN tensor with entry [:, i] is the xyz world-space coordinate corresponding to the camera-space
+    coordinate camera_space_xyzs[:, i]
+    """
+    # Adapted from https://github.com/devendrachaplot/Neural-SLAM.
+
+    # First compute the transformation that points undergo
+    # due to the camera's horizon
+    psi = -horizon * np.pi / 180
+    cos_psi = np.cos(psi)
+    sin_psi = np.sin(psi)
+    # fmt: off
+    horizon_transform = camera_space_xyzs.new(
+        [
+            [1, 0, 0], # unchanged
+            [0, cos_psi, sin_psi],
+            [0, -sin_psi, cos_psi,],
+        ],
+    )
+    # fmt: on
+
+    # Next compute the transformation that points undergo
+    # due to the agent's rotation about the y-axis
+    phi = -rotation * np.pi / 180
+    cos_phi = np.cos(phi)
+    sin_phi = np.sin(phi)
+    # fmt: off
+    rotation_transform = camera_space_xyzs.new(
+        [
+            [cos_phi, 0, -sin_phi],
+            [0, 1, 0], # unchanged
+            [sin_phi, 0, cos_phi],],
+    )
+    # fmt: on
+
+    # Apply the above transformations
+    view_points = (rotation_transform @ horizon_transform) @ camera_space_xyzs
```

Comment:
Didn't check for the correctness of the signs in the rotation matrices, but LGTM.